packages/react/src/primitives/PhoneNumberField/PhoneNumberField.tsx

```diff
@@ -28,10 +29,40 @@ const PhoneNumberFieldPrimitive: Primitive<PhoneNumberFieldProps, 'input'> = (
     onInput,
     size,
     variation,
+    defaultValue,
+    onChange,
+    value,
     ...rest
   },
   ref
 ) => {
+  // state for the current dial code and default to the defaultDialCode
+  const [dialCode, setDialCode] = React.useState<string>(
+    (defaultDialCode as string) ?? ''
+  );
+  // state for the formatted phone number and default to the defaultValue
+  const [formattedPhoneNumber, setFormattedPhoneNumber] =
+    React.useState<string>(
+      formatPhoneNumber(dialCode, (value as string) ?? defaultValue ?? '')
+    );
+  const onNumberChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const phoneNumber = event.target.value;
+    // format the phone number with the current dial code
+    const formatted = formatPhoneNumber(dialCode, phoneNumber);
+    setFormattedPhoneNumber(formatted);
+    onChange?.(event);
+  };
+  const onDialCodeChangeWrapper = (
+    event: React.ChangeEvent<HTMLSelectElement>
+  ) => {
+    const code = event.target.value;
+    // Call the onDialCodeChange prop if it exists
+    setDialCode(code);
+    // Reformat the phone number with the new dial code
+    const formatted = formatPhoneNumber(code, formattedPhoneNumber);
+    setFormattedPhoneNumber(formatted);
```

Comment:
Will this be formatted on every keystroke? Can you share a GIF/video showing the UX?

Reply:
Sure, see the attached.

https://github.com/aws-amplify/amplify-ui/assets/19784835/756a19b8-27a5-4bbf-b8ec-8e5a58df7ccc

Reply:
Do we have some default debounce logic here? I would have thought the same thing as Scott

Reply:
I believe the debounced logic means no on the fly formatting. The formatting will happen after the debounced, so if users type fast it will happen after they stop typing. 
Could you please provide more details about the desired debounce logic here?

Reply:
Is there debouncing here? Don't see anything added in this PR, nor do i believe there is debouncing in the existing in the existing code

Reply:
Think we are trying to clarify if there is debouncing happening in your example video, not requesting it. The lack of formatting until you have completed typing gives the appearance that there is some form of debounce happening, can you just clarify if the example video includes debouncing or not?

Reply:
There is no debouncing, the phone number gets formatted after being fully typed. I do agree that it might be not intuitive for the user to wait until the end to see the number formatted. I can work on updating the logic so it would be formatted on the fly while you're typing, to make the formatting more intuitive and interactive, if it makes sense

Reply:
Would expect the formatting on each change event

Reply:
Added the formatting on each change event